src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventPipeEventDispatcher.cs

```diff
@@ -142,12 +142,13 @@ private void SetStopDispatchTask()
         {
             Debug.Assert(Monitor.IsEntered(m_dispatchControlLock));
 
-            if (m_dispatchTask != null)
+            if (m_dispatchTaskCancellationSource?.IsCancellationRequested ?? true)
             {
-                Debug.Assert(m_dispatchTaskCancellationSource != null);
-                m_dispatchTaskCancellationSource?.Cancel();
-                EventPipeInternal.SignalSession(m_sessionID);
+                return;
             }
+
+            m_dispatchTaskCancellationSource.Cancel();
+            EventPipeInternal.SignalSession(m_sessionID);
```

Comment:
Since we have handed of m_sessionID to the dispatch loop and this is the last thing we should do with it (will be freed by dispatch loop past this)it might make sense to set it to 0 to prevent any future use of it, since after this point that session needs to be considered freed.

Reply:
Another note, this pattern builds on the fact that the dispatch loop plays nice (maybe it always will) and not stopped before we get this point, or m_sessionID might actually reference another session. There is also a race between the dispatch loop (since its not locking m_dispatchControllLock) and this method, between call to cancel and following call to SIgnalSession, if the dispatch loop gets to EventPipeInternal.Disable(sessionID) before SignalSession happens, we will still signal a freed session (or reused different session). That could be fixed by making sure the dispatch loop grabs the m_dispatchControllLock before calling EventPipeInternal.Disable(sessionID).

Reply:
We see races happening due to hardware optimizations and memory reordering between individual instructions, so if there is a race it will sooner or later happen (or at least bother us when investigating other issue since we know its there).